### PR TITLE
Use local POST request for gestures to prevent touch lockup

### LINKED ISSUE #21263: Some images can disable touch inputs
### Checked for duplicates?

- [x] This issue is not a duplicate

### Does it also happen in the desktop version?

- [x] This bug does not occur in the latest version of Anki Desktop

### What are the steps to reproduce this bug?

1. Download [no_touch_events.apkg.zip](https://github.com/user-attachments/files/28893064/no_touch_events.apkg.zip).
2. Rename it to `no_touch_events.apkg` and import it into AnkiDroid.
3. Open the deck in the new study screen.
4. Show the answer.
5. Double tap on the image but don't lift your finger after the second tap. Instead, wait for a few seconds.
6. Do step 5 again.

Now you can't use any gestures and the HTML elements are not interactive (`:hover` doesn't work, links don't work, etc).

https://github.com/user-attachments/assets/7e6764da-2268-4ba0-a451-4011f2f68036

### Expected behaviour

Touch inputs should work.

### Debug info

```text
AnkiDroid Version = 2.25.0alpha1 (98bddabfa055eb61e0b5f5ab2dfde43f3f174ee6)  
Backend Version = 0.1.64-anki25.09.2 (25.09.2 3890e12c9e48c028c3f12aa58cb64bd9f8895e30)  
Android Version = 13 (SDK 33)  
ProductFlavor = full  
Device Info = Xiaomi | POCO | mondrian | mondrian_global | 23013PC75G | qcom  
WebView Info = [com.google.android.webview | 782709103]: Mozilla/5.0 (Linux; Android 13; 23013PC75G Build/TKQ1.220905.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/149.0.7827.91 Mobile Safari/537.36  
ACRA UUID = 4afc3b3f-e93f-4a66-bc71-2b9476ef9eb0  
FSRS = 5.1.0 (Enabled: true)  
Crash Reports Enabled = true
```

### (Optional) Anything else you want to share?

- This bug happens in the new study screen only.
- This bug also happens in 2.24.0 and 2.23.3.
- Some images don't cause this issue. It seems like only images with about the same aspect ratio as the one above are problematic.

Similar issues: https://github.com/ankidroid/Anki-Android/issues/21045 and https://github.com/ankidroid/Anki-Android/issues/19165.

### Research

- [x] I have checked the [manual](https://docs.ankidroid.org/) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by DoomsCoder**
## Purpose / Description
A bug in the new study screen where double tapping and holding on images causes the Webview to lockup and ignore all gesture and links.

## Fixes
* Fixes #21263 

## Approach
I replaced the `window.location.href` gesture tracking with the existing `localRequest` API to send gesture payloads via POST request. This keeps the dynamic template sandbox completely secure, while preventing the WebView from thinking a new webpage is loading, which stops the Chromium engine from locking up its touch sensors. I also added a `touchcancel` listener to reset the gesture state when zooming.

## How Has This Been Tested?

https://github.com/user-attachments/assets/0ee6d6cc-e4e3-433e-98dd-3d4ddffb9305

## Debug Info
```
AnkiDroid Version = 2.25.0alpha4-debug (d9ee03d66680ee691f8760f8dda05fd11acd06e1)  
Backend Version = 0.1.66-anki26.05 (26.05 e64c6b1aee3e8d668fb8bbe084beada8e070d985)  
Android Version = 13 (SDK 33)  
ProductFlavor = play  
Device Info = motorola | motorola | rhodei | rhodei_g | moto g62 5G | qcom  
WebView Info = [com.google.android.webview | 797706403]: Mozilla/5.0 (Linux; Android 13; moto g62 5G Build/T1SSIS33.1-75-7-16; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/152.0.7977.64 Mobile Safari/537.36  
ACRA UUID = aa511b45-fc15-4f87-a9d1-042ac9e9631c  
FSRS = 5.2.0 (Enabled: false)  
Crash Reports Enabled = false

```
## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by BrayanDSO** [comment]
Using `JavascriptInterface` with dynamic content provided by card content/card templates isn't safe, as far as I understand.

Is there another approach to fix the issue?

**Comment by DoomsCoder** [comment]
@BrayanDSO My bad! I didnt think about the security issue.

i did some research what if we reuse `localRequest()` from js instead ? we can send the gesture payload via POST and intercept it in `ReviewerViewModel.handlePostRequest()` exactly like we do for `focusin`.

It keeps the security completely safe and prevents the webview lockup. Let me know if you like the approach!

**Comment by BrayanDSO** [comment]
Using POST requests is okay. Go for it

File: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerViewModel.kt [hunk 1]
```diff
@@ -445,6 +449,32 @@ class ReviewerViewModel(
                 onStateMutationCallback()
                 return byteArrayOf()
             }
+            "multiFingerTap" -> {
+                val json = JSONObject(String(bytes))
+                val touchCount = json.getInt("touchCount")
+                val gestureUri = "gesture://multiFingerTap/?touchCount=$touchCount".toUri()
+                gestureUriFlow.emit(gestureUri)
+                return byteArrayOf()
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerViewModel.kt:452]: if we have communication between the ViewModel and the View, why keep using an URI to do that?

URIs were useful when we communicated directly from the JavaScript code to the View, but they don't see necessary anymore.

Replacing URIs require a bigger refactor, but it's an improvement over the old code.

Comment (DoomsCoder) [reply AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerViewModel.kt:452]: Makes sense! I will refactor GestureParser and other things to accept a Kotlin data class instead of URI.

**Review by david-allison — CHANGES_REQUESTED** [review]
This is broken on the new study screen.

Please test more, and ideally add a regression test (a minimum of `@NeedsTest`) so this doesn't occur in future.

**Comment by DoomsCoder** [comment]
@david-allison I have tested the gesture on mobile device and they are working perfectly now. I have also added the `@NeedsTest`.Please let me know if i missed anything.

File: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerFragment.kt [hunk 2]
```diff
@@ -690,12 +705,20 @@ class ReviewerFragment :
...
+                            RawGesture.TapOrSwipe(
+                                x = url.getQueryParameter("x")?.toIntOrNull() ?: 0,
+                                y = url.getQueryParameter("y")?.toIntOrNull() ?: 0,
+                                deltaX = url.getQueryParameter("deltaX")?.toIntOrNull() ?: 0,
+                                deltaY = url.getQueryParameter("deltaY")?.toIntOrNull() ?: 0,
+                                time = url.getQueryParameter("time")?.toLongOrNull() ?: 0,
+                                scrollDirection = url.getQueryParameter("scrollDirection"),
+                            )
+                        }
+                    handleGesture(rawGesture)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerFragment.kt]: URLs aren't being used anymore, so why they are being handled here?

File: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerViewModel.kt [hunk 3]
```diff
@@ -445,6 +448,27 @@ class ReviewerViewModel(
...
+                val json = JSONObject(String(bytes))
+                val rawGesture =
+                    RawGesture.TapOrSwipe(
+                        x = json.getInt("x"),
+                        y = json.getInt("y"),
+                        deltaX = json.getInt("deltaX"),
+                        deltaY = json.getInt("deltaY"),
+                        time = json.getLong("time"),
+                        scrollDirection = if (json.has("scrollDirection")) json.getString("scrollDirection") else null,
+                    )
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerViewModel.kt]: add a constructor into `TapOrSwipe` that takes bytes as arguments, then add a test to check if the parsing is correct

File: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerFragment.kt (hunk 2 above)
Comment (DoomsCoder) [reply AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/ReviewerFragment.kt]: Removed the gesture url fallback from the `handleUrl`

File: AnkiDroid/src/main/assets/scripts/ankidroid-reviewer.js [hunk 4]
```diff
@@ -114,7 +114,11 @@ globalThis.ankidroid.onTypeAnswerKeyDown = function (event) {
                 if (Date.now() - touchStartTime > MULTI_TOUCH_TIMEOUT) {
                     return;
                 }
-                window.location.href = `${SCHEME}://multiFingerTap/?touchCount=${touchCount}`;
+
+                fetch(`/ankidroid/multiFingerTap`, {
+                    method: "POST",
+                    body: JSON.stringify({ touchCount: touchCount }),
+                });
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/assets/scripts/ankidroid-reviewer.js]: extract `localRequest` and reuse it here and in the other fetch you've added

Comment (DoomsCoder) [reply AnkiDroid/src/main/assets/scripts/ankidroid-reviewer.js]: done

File: AnkiDroid/src/main/assets/scripts/ankidroid-reviewer.js [hunk 5]
```diff
@@ -31,21 +31,25 @@ globalThis.ankidroid.onTypeAnswerKeyDown = function (event) {
     }
 };
 
+// ============================================================================
+// Helper functions
+// ============================================================================
+
+function localRequest(endpoint, payload = {}) {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/assets/scripts/ankidroid-reviewer.js]: the method shouldn't be accessible to the user.

Put both the input focus listeners and the gesture parsing code under the same IIFE block

File: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/GestureParser.kt [hunk 6]
```diff
@@ -27,9 +27,37 @@ import kotlinx.coroutines.CoroutineScope
...
 import kotlinx.coroutines.launch
+import org.json.JSONObject
 import timber.log.Timber
 import kotlin.math.abs
 
+sealed class RawGesture {
+    data class MultiTouch(
+        val touchCount: Int,
+    ) : RawGesture() {
+        constructor(bytes: ByteArray) : this(JSONObject(String(bytes)).getInt("touchCount"))
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/reviewer/GestureParser.kt]: add a JSONObject constructor to this class as well

**Comment by DoomsCoder** [comment]
@david-allison @BrayanDSO I have made the changes you asked for. Please let me know if everything looks good.
